checkpointBuildTools.prepareCheckpointBuild: stop at `installPhase`

ShamrockLee opened the pull request:
## Description of changes

Switch off phases following `installPhase`, including `fixupPhase`, `installCheckPhase`, `distPhase` and `postPhases` of the checkpoint build artifact.

Add `postInstallCheck` to `hello` to test the existence of main program and to serve as a better test target of `tests.checkpointBuildTools`.

Cc: @messemar

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ShamrockLee committed:
checkpointBuildTools.prepareCheckpointBuild: stop at install

Switch off phases following installPhase, including fixupPhase,
installCheckPhase, distPhase and postPhases.

ShamrockLee commented:
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

ShamrockLee commented:
@ofborg build tests.checkpointBuildTools

messemar reviewed (changes requested):
Thank you for the contribution!

I left some thoughts.

From my perspective,  the change looks reasonable because, in the prepare step,  all phases after the install phase operate on non meaningful data or even garbage.

@roberth can you have a look here too ?

ShamrockLee committed:
hello: add postInstallCheck

Set doInstallCheck = true.
Test the existence of main program in postInstallCheck.

ShamrockLee commented on pkgs/test/checkpointBuild/default.nix:
```diff
@@ -1,12 +1,24 @@
+    '';
+  });
+  baseHelloArtifacts = checkpointBuildTools.prepareCheckpointBuild hello-with-installcheck;
+  patchedHello = hello-with-installcheck.overrideAttrs (old: {
```
It's not easy to find a simple-enough package that has `fixupPhase` or `installCheckPhase`. Let's give `hello` a `postInstallCheck` phase then.

messemar replied:
Well, then lets keep the constructed sample in the tests. There is no reason to change the derivation of hello IMHO.

roberth commented on pkgs/build-support/checkpoint-build.nix:
```diff
@@ -53,7 +53,12 @@ rec {
       mkdir -p $out/outputs
       cp -r ./* $out/outputs/
       runHook postCheckpointInstall
+      unset postPhases
```
Should we also unset `preInstallPhases`, `preFixupPhases` and `preDistPhases`?

messemar replied:
Sounds good. `preInstallPhases` has no effect, because the Phase is already over, but the other ones can be unset

roberth reviewed (commented):
Looks like these phases were simply forgotten in the initial implementation. This looks more than sensible.

ShamrockLee commented:
@ofborg build tests.checkpointBuildTools

ShamrockLee commented:
Result of `nixpkgs-review pr 295594` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>hello (tests.config.allowPkgsInPermittedInsecurePackages)</li>
    <li>tests.checkpointBuildTools</li>
    <li>tests.pkgs-lib</li>
    <li>tests.testers.nixosTest-example</li>
    <li>tests.testers.runNixOSTest-example</li>
    <li>tests.testers.testBuildFailure.helloDoesNotFail</li>
    <li>tests.trivial-builders.concat</li>
    <li>tests.trivial-builders.linkFarm</li>
    <li>tests.trivial-builders.references</li>
    <li>tests.trivial-builders.writeShellApplication</li>
    <li>tests.trivial-builders.writeStringReferencesToFile</li>
  </ul>
</details>

ConnorBaker reviewed (approved):
Fantastic -- thank you for your work on this!
